runtime/shared_heap.c

```diff
@@ -401,16 +395,35 @@ static void* pool_allocate(struct caml_heap_state* local, sizeclass sz) {
   if (!r) return 0;
 
   p = r->next_obj;
-  next = (value*)p[1];
+  /* assert that p is inside the pool */
+  CAMLassert(p >= (value*)r + POOL_HEADER_WSIZE && p < (value*)r + POOL_WSIZE);
```

Comment:
Similarly, this would better be written as two distinct `CAMLassert`, one for each condition.

Reply:
Thanks, resolved in https://github.com/ocaml/ocaml/pull/13616/commits/912f3665b7ced848a62a616203b3c0c9d18db391